Source/Core/DolphinQt/SkylanderPortal/SkylanderPortalWindow.cpp

```diff
@@ -0,0 +1,867 @@
+{
+  Config::SetBaseOrCurrent(Config::MAIN_EMULATE_SKYLANDER_PORTAL, emulate);
+  m_group_skylanders->setVisible(emulate);
+}
+
+void SkylanderPortalWindow::CreateSkylander(u8 slot)
+{
+  CreateSkylanderDialog create_dlg(this);
+  if (create_dlg.exec() == CreateSkylanderDialog::Accepted)
+  {
+    LoadSkylanderPath(slot, create_dlg.GetFilePath());
+  }
+}
+
+void SkylanderPortalWindow::LoadSkylander(u8 slot)
+{
+  const QString file_path = DolphinFileDialog::getOpenFileName(
+      this, tr("Select Skylander File"), s_last_skylander_path, tr("Skylander (*.sky);;"));
+  if (file_path.isEmpty())
+  {
+    return;
+  }
+  s_last_skylander_path = QFileInfo(file_path).absolutePath() + tr("/");
+
+  LoadSkylanderPath(slot, file_path);
+}
+
+void SkylanderPortalWindow::LoadSkylanderPath(u8 slot, const QString& path)
+{
+  File::IOFile sky_file(path.toStdString(), "r+b");
+  if (!sky_file)
+  {
+    QMessageBox::warning(
+        this, tr("Failed to open the skylander file!"),
+        tr("Failed to open the skylander file(%1)!\nFile may already be in use on the portal.")
+            .arg(path),
+        QMessageBox::Ok);
+    return;
+  }
+  std::array<u8, 0x40 * 0x10> file_data;
+  if (!sky_file.ReadBytes(file_data.data(), file_data.size()))
+  {
+    QMessageBox::warning(
+        this, tr("Failed to read the skylander file!"),
+        tr("Failed to read the skylander file(%1)!\nFile was too small.").arg(path),
+        QMessageBox::Ok);
+    return;
+  }
+
+  ClearSkylander(slot);
+
+  u16 sky_id = file_data[0x11];
+  u16 sky_var = file_data[0x1D];
+  sky_id <<= 8;
+  sky_var <<= 8;
+  sky_id |= file_data[0x10];
+  sky_var |= file_data[0x1C];
+
+  DEBUG_LOG_FMT(IOS_USB, "Sky Id: {}, 0x10: {} 0x11: {}", sky_id, file_data[0x10], file_data[0x11]);
+  DEBUG_LOG_FMT(IOS_USB, "Sky Var: {}, 0x1D: {} 0x1C: {}", sky_var, file_data[0x1D],
+                file_data[0x1C]);
+
+  auto& system = Core::System::GetInstance();
+
+  u8 portal_slot = system.GetSkylanderPortal().LoadSkylander(file_data.data(), std::move(sky_file));
+  if (portal_slot == 0xFF)
+  {
+    QMessageBox::warning(this, tr("Failed to load the skylander file!"),
+                         tr("Failed to load the skylander file(%1)!\n").arg(path), QMessageBox::Ok);
+    return;
+  }
+  m_sky_slots[slot] = {portal_slot, sky_id, sky_var};
+  UpdateEdits();
+}
+
+void SkylanderPortalWindow::ClearSkylander(u8 slot)
+{
+  auto& system = Core::System::GetInstance();
+  if (auto slot_infos = m_sky_slots[slot])
+  {
+    system.GetSkylanderPortal().RemoveSkylander(slot_infos->portal_slot);
+    m_sky_slots[slot].reset();
+    UpdateEdits();
+  }
+}
+
+void SkylanderPortalWindow::UpdateEdits()
+{
+  for (auto i = 0; i < MAX_SKYLANDERS; i++)
+  {
+    QString display_string;
+    if (auto sd = m_sky_slots[i])
+    {
+      auto found_sky = list_skylanders.find(std::make_pair(sd->sky_id, sd->sky_var));
+      if (found_sky != list_skylanders.end())
+      {
+        display_string = tr(found_sky->second);
```

Comment:
If you want this to work, you'll also need to mark each string in the definition of `list_skylanders` as translatable, for example:

```cpp
    {{0, 0x0000}, _trans("Whirlwind")},
```

(@JosJuice, do you know if it's preferable to use _trans or QT_TR_NOOP for this? Why do we have both macros anyway?)

Reply:
We have `QT_TR_NOOP` because Qt defines it, and we have `_trans` because we can't use `QT_TR_NOOP` in core code (since it would be a Qt dependency).

As long as you are able to include the relevant header, I don't think it really matters which one you use. Normally we use `QT_TR_NOOP` in DolphinQt, but I suppose you could make the argument that since it's not a `QString`, using `_trans` would make sense... I don't have a strong opinion on this one.

Reply:
Now that I think about it, this big set of strings is probably going to be pretty confusing for translators since they won't be grouped in a way that would make it easy for the translators to realize that they're Skylanders figures. (Some of them like "Bash" or "Camo" would without context probably not be treated as names...)

First of all: Do these figures actually have different names in different languages?

If so, maybe we could specify a context for these strings. You would do this like `QT_TRANSLATE_NOOP("Skylanders", "Whirlwind")` in place of `QT_TR_NOOP("Whirlwind")`, with the "Skylanders" parameter being treated as the *context*. You would then also have to specify the context when calling `tr`.

Reply:
I'm fairly certain that the names aren't translated in other languages - let me just check to see though

Reply:
Doesn't look they have translations, will changes these back to non translatable Strings